### DISCUSSION TITLE
how to handle direct response from middlware while try server actions

### DISCUSSION THREAD

**Discussion by YellaGoya** [vercel/next.js#62446] [Help] [answered]
### Summary

I implemented a simple rate limit through middleware and have the middleware respond directly with a 429 response when the limit is exceeded. I can see the log of the 429 response in the browser's console, but it is not caught in the try / catch statement.

If someone could give me a hint on this, I would be very grateful.

next.js 14 / app router

### middleware.ts
```
export const middleware = (req: NextRequest) => {
  let res = NextResponse.next();

  if (req.nextUrl.pathname.startsWith('/setting') && req.method === 'POST') {
    const forwardedFor = headers().get('x-forwarded-for');
    const ip = forwardedFor ? forwardedFor.split(':')[0] : headers().get('x-real-ip');

    if (!ip) {
      return res;
    }

    const now = Date.now();
    const windowMs = 60 * 1000;
    const blockDuration = 15 * 60 * 1000;
    const max = 5;

    if (!store[ip] || (now > store[ip].timestamp && now > store[ip].blockUntil)) {
      store[ip] = { count: 1, timestamp: now + windowMs, blockUntil: 0 };
    } else if (store[ip].count < max) {
      store[ip].count++;
    } else {
      store[ip].blockUntil = now + blockDuration;
      res = NextResponse.json({ error: 'Too Many Requests' }, { status: 429 });
    }
  }

  return res;
};
```

### component.js
```
  const loginDispatch = async (form) => {
    try {
      setResult(await authenticate(form));
      console.log('no problem here.');
    } catch {
      console.log('error occured.');
    }
  };
```

`await authenticate(form)` returns `undefined`

### browser console log
![image](https://github.com/vercel/next.js/assets/83230053/313bbf93-0b8a-47b2-8566-d75cda992dff)

**Comment by icyJoseph** [comment]
Hi,

Is that using the browser's `fetch`? Or I guess, React is using it under the hood to make the request. 

The thing with fetch is that it doesn't throw on 4xx range. 

Let me investigate a bit more, but it is most likely because of how fetch doesn't throw. 

In the meantime here's one resource to read up:

- https://web.dev/articles/fetch-api-error-handling

**Reply by YellaGoya** [reply]
Thanks for the reply! I don't use the browser's fetch, I use the server api using the server actions provided by next.js 14 app router. Since I just import and use the api method declared on the server side, I can't get NextResponse like fetch, so it seems to return undefined without catching it.

In the latest next.js docs and tutorials, the server api is used via server actions, should I use a different method?

**Reply by YellaGoya** [reply]
### lib/action/user.ts
```
'use server';

...

export const authenticate = async (prevState: string | undefined, formData: FormData) => {
  try {
    const loginValuesObj = Object.fromEntries(formData.entries());

    await signIn('credentials', { ...loginValuesObj, redirect: false });

    return { success: true };
  } catch (error) {
    console.log(error);
    if (error instanceof AuthError) {
      switch (error.type) {
        case 'CredentialsSignin':
          return { success: false, error: '( ! ) 잘못된 태그 정보 입니다.' };
        default:
          return { success: false, error: '( ! ) 연결이 불안정합니다. 다시 시도 해주세요.' };
      }
    }
  }
};
```

### components/login-form.jsx
```
'use client';

...
import { authenticate } from 'lib/action/user';
...

const LoginForm = ({ backward, forward }) => {
  const [result, setResult] = useState({ success: null, error: null });

  const loginDispatch = async (form) => {
    try {
      setResult(await authenticate(form));
      // console.log('no problem here.');
    } catch {
      console.log('error occured.');
    }
  };
  
  ...
```

**Reply by YellaGoya** [reply]
And if the middleware returns anything else with NextResponse.json( ... ) before the server-side API is executed, the API will never be executed. It will only run when passed to NextResponse.next().

If the middleware throws an error, it will change to
500 internal error, it will not be caught.

**Reply by icyJoseph** [reply]
In the network dev tools, if you inspect the 429, does the payload show the data?

**Reply by icyJoseph** [reply]
Yeah I see that the Server Action does use `fetch` though, I don't have time to debug this though, but it is something with how the server action is processed.

**Reply by YellaGoya** [reply]
If the middleware does not send back NextResponse.json ( ... ) to notify rate limit has been exceeded, it will work fine.

However, when responding with NextResponse.json( ... ), the server actions method in next.js doesn't seem to handle simple responses.

![image](https://github.com/vercel/next.js/assets/83230053/d8fd4863-d0bf-4866-992c-e5c4622aa518)

![image](https://github.com/vercel/next.js/assets/83230053/7e968e98-0e28-4f5e-8732-562a625bd002)

**Reply by icyJoseph** [reply]
Hi,

It is here that the your issue is happening:

- https://github.com/vercel/next.js/blob/ae524fb24499fb8caf5b70eb8ce4cc96a5301565/packages/next/src/client/components/router-reducer/reducers/server-action-reducer.ts#L54

From that's visible there, the `ok` status of the response is not checked, and the response is processed for other purposes.

**Reply by icyJoseph** [reply, marked as answer]
Looking a bit deeper it feels like this goes back to react-dom-server as well :/ I'd say open an issue. There is people who can answer this better than I can over there, since this is a rather tricky use case.

All you need is a middleware that does this:

```ts
import { NextRequest, NextResponse } from "next/server";

export function middleware(req: NextRequest) {
  if (req.method === "POST") {
    return Response.json(
      { message: "Hello" },
      {
        status: 429,
      }
    );
  }

  return NextResponse.next();
}

```

And then any server action activated from the client will trigger it (because of the POST method). 

Say:

```jsx
"use client";

import { foo } from "./_action";

export default function Home() {
  const sendForm = async () => {
    const data = new FormData();

    data.append("foo", "bar");

    const result = await foo(data);
  };

  return <button onClick={sendForm}>Send it</button>;
}

```

Where foo is:

```ts
"use server";

export async function foo(data: FormData) {
  console.log(data.entries());
  return { foo: "bar" };
}
```

Maybe the server action needs to check for response.ok, and throw in that case? or if the content type is application/json, or if there's a `x-middleware-response` header, maybe take that, idk.

**Reply by YellaGoya** [reply]
you are my angel...

**Reply by boncz92** [reply]
Just curious, what did you end up doing to solve your issue? Im also trying to return a failed auth from middleware when routing to server action and its causing some issues.

**Reply by boncz92** [reply]
I would rather not include a function at the start of all server actions to ensure that there isnt a header that indicates that authentication failed lol.
